opm/autodiff/StandardWellsDense_impl.hpp

```diff
@@ -2539,16 +2619,268 @@ namespace Opm {
+    computeWellPotentialWithTHP(const Simulator& ebosSimulator,
+                                const int well_index,
+                                const double initial_bhp, // bhp from BHP constraints
+                                const std::vector<double>& initial_potential) const
+    {
+        // TODO: pay attention to the situation that finally the potential is calculated based on the bhp control
+        // TODO: should we consider the bhp constraints during the iterative process?
+        const int np = wells().number_of_phases;
+
+        assert( np == int(initial_potential.size()) );
+
+        std::vector<double> potentials = initial_potential;
+        std::vector<double> old_potentials = potentials; // keeping track of the old potentials
+
+        double bhp = initial_bhp;
+        double old_bhp = bhp;
+
+        bool converged = false;
+        const int max_iteration = 1000;
+        const double bhp_tolerance = 1000.; // 1000 pascal
+
+        int iteration = 0;
+
+        while ( !converged && iteration < max_iteration ) {
+            bhp = initial_bhp;
+
+            // the well controls
+            const WellControls* well_control = wells().ctrls[well_index];
+            // The number of the well controls/constraints
+            const int nwc = well_controls_get_num(well_control);
+
+            for (int ctrl_index = 0; ctrl_index < nwc; ++ctrl_index) {
+                if (well_controls_iget_type(well_control, ctrl_index) == THP) {
+                    double aqua = 0.0;
+                    double liquid = 0.0;
+                    double vapour = 0.0;
+
+                    const Opm::PhaseUsage& pu = phase_usage_;
+
+                    if (active_[ Water ]) {
+                        aqua = potentials[pu.phase_pos[ Water ] ];
+                    }
+                    if (active_[ Oil ]) {
+                        liquid = potentials[pu.phase_pos[ Oil ] ];
+                    }
+                    if (active_[ Gas ]) {
+                        vapour = potentials[pu.phase_pos[ Gas ] ];
+                    }
+
+                    const int vfp        = well_controls_iget_vfp(well_control, ctrl_index);
+                    const double& thp    = well_controls_iget_target(well_control, ctrl_index);
+                    const double& alq    = well_controls_iget_alq(well_control, ctrl_index);
+
+                    // Calculating the BHP value based on THP
+                    const int first_perf = wells().well_connpos[well_index]; //first perforation
+
+                    const WellType& well_type = wells().type[well_index];
+                    if (well_type == INJECTOR) {
+                        const double dp = wellhelpers::computeHydrostaticCorrection(
+                                          wells(), well_index, vfp_properties_->getInj()->getTable(vfp)->getDatumDepth(),
+                                          wellPerforationDensities()[first_perf], gravity_);
```

Comment:
Ok for now, but add TODO comment to check if it is always correct to use first_perf here.